rpc/wrtc_base_stream.go

```diff
@@ -104,6 +113,15 @@ func (s *webrtcBaseStream) RecvMsg(m interface{}) error {
 			return proto.Unmarshal(msgBytes, m.(proto.Message))
 		}
 		return s.ctx.Err()
+	case <-s.channelCtx.Done():
+		msgBytes, err := checkLastOrErr(s.channelCtx.Err())
+		if err != nil {
+			return err
+		}
+		if msgBytes != nil {
+			return proto.Unmarshal(msgBytes, m.(proto.Message))
+		}
+		return s.channelCtx.Err()
```

Comment:
This code is redundant with the case above, but `fallthrough` does not exist for select, and you cannot have multiple `case` values for a single logical branch. Abstracting into a function felt confusing with the pre-existing `checkLastOrErr` inline function.

Reply:
I think this is fine the way it is.

`checkLastOrErr` confuses me. It feels like an optimization. Like if we hit the timeout case, we should do one last check if we got any messages. If it's for correctness, I wish we knew why so we could document. If it's an optimization, I'm unconvinced it's helpful and removing it would elide us from worrying about this redundancy.

Can keep it in mind to poke at these code paths more the next time we have a reason to look at this area.

Reply:
>  Like if we hit the timeout case, we should do one last check if we got any messages.

That's how I understand it, too. If we fall into a context error branch, we see whether the msgCh contained a message, too. If it did, we throw that message into the `m` output parameter. If it didn't, we return the context error. If the channel was already closed, we try to return the last error stored on the stream, replacing an internal expected closure with `io.EOF`. The one branch I don't really understand is:
```
if origErr == nil {
   return nil, io.EOF
}
```
I don't know how we could ever fall into that case if the context is done and `ctx.Err()` was passed in as `origErr`.

In any case, I also can't describe the "correctness" that function is trying to enforce, and it feels more like an attempt to respect actual errors on the stream over context errors when returning an error to the user (hoping that actual errors are more informative/relevant than context errors). I've left a comment in `checkLastOrErr` to explain my understanding; let me know what you think.

Reply:
[q] I'm totally happy with the current implementation, but just out of curiosity - is there a way to "merge" contexts in golang? not sure if that's even appropriate here, but say we did end up wanting to react to both contexts here the same way?

Reply:
Another good q. There's no obvious function/method I know of to just "merge" two contexts, but there are two options I can think of that might be more elegant than storing two contexts on the same struct. The first method is often referred to as a "fan-in" of channels (see [this old article](https://go.dev/blog/pipelines) on go concurrency patterns); we _could_ fan-in the `Done` channels of the two contexts and check the aggregated channel. The second method is establishing a parent/child relationship between the channel context and the stream context. That way, the stream context's `Done` channel will receive a value when the channel context's does (but _not_ vice-versa). I'll try the latter method locally and see if it cleans things up.

Reply:
> There's no obvious function/method I know of to just "merge" two contexts

...yes [there is](https://github.com/viamrobotics/goutils/blob/main/context.go#L12-L25) haha; it's in this repo. Things seem cleaner using `MergeContext*`, and the hang still disappears in my repro, so I'll use those methods. FWIW they use the "fan-in" method mentioned above. Thanks for the idea!

I also discovered that webRTC _server_ streams seem to already incorporate the server channel's context into the context passed to `RecvMsg` [here](https://github.com/viamrobotics/goutils/blob/main/rpc/wrtc_server_channel.go#L121). So, I don't think we need any extra server logic for server-side `RecvMsg` hangs.